### PR TITLE
o/snapstate: assign properly local revisions and be more careful when removing components

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
* Look at all sequence points to assign a local revision. We want it to be unique across all sequence points, as otherwise we would be duplicating file names.
* Look at all the sequence to find out if we can remove a component container when we are replacing it, as it might be used still in a non-current sequence point to where we might want to revert eventually.

### CODE REVIEW

File: overlord/snapstate/snapmgr.go [hunk 1]
```diff
@@ -399,6 +399,37 @@ func (snapst *SnapState) IsComponentInCurrentSeq(cref naming.ComponentRef) bool
...
+			continue
+		}
+		compInSeqPt := seqPt.FindComponent(cref)
+		if compInSeqPt == nil {
+			continue
+		}
+		if compInSeqPt.SideInfo.Revision == currentComp.SideInfo.Revision {
+			return true
+		}
+	}
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:428]: is some of it maybe a SnapSequence helper?

File: overlord/snapstate/snapmgr.go [hunk 2]
```diff
@@ -411,6 +442,25 @@ func (snapst *SnapState) LocalRevision() snap.Revision {
 	return local
 }
 
+// LocalComponentRevision returns the "latest" local revision for the compName
+// component. Local revisions start at -1 and are counted down. 0 will be
+// returned if no local revision for the component is found.
+func (snapst *SnapState) LocalComponentRevision(compName string) snap.Revision {
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:448]: same question

**Review by pedronis — COMMENTED** [review]
thank you, I wonder whether part of the new helpers belongs to methods on SnapSequence

File: overlord/snapstate/snapmgr.go (hunk 2 above)
Comment (andrewphelpsj) [inline overlord/snapstate/snapmgr.go:448]: The name of this function could maybe better imply it is "latest" revision of the local component. Reading the name now, it isn't clear that it is looking for the latest local component.

File: overlord/snapstate/snapstate_test.go [hunk 3]
```diff
@@ -5288,6 +5288,78 @@ func (s *snapmgrTestSuite) TestSnapStateLocalRevision(c *C) {
 	c.Assert(snapst.LocalRevision().Unset(), Equals, true)
 }
 
+func (s *snapmgrTestSuite) TestSnapStateLocalComponentRevision(c *C) {
+	si7 := snap.SideInfo{
+		RealName: "some-snap",
+		Revision: snap.R(-7),
+	}
+	csi1 := snap.NewComponentSideInfo(naming.NewComponentRef("some-snap", "mycomp"), snap.R(-3))
```
Comment (andrewphelpsj) [inline overlord/snapstate/snapstate_test.go:5296]: Maybe put a non-local revision of the components in here, just to make sure that we ignore those?

File: overlord/snapstate/snapmgr.go (hunk 2 above)
Comment (andrewphelpsj) [reply overlord/snapstate/snapmgr.go:448]: Although I see now that this matches the pattern of the preexisting `SnapState.LocalRevision`, so matching that is probably more important.

**Review by andrewphelpsj — APPROVED** [review]
Looks good, just a couple comments.

File: overlord/snapstate/snapmgr.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go]: I have moved most of this to the sequence package.

File: overlord/snapstate/snapmgr.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapmgr.go:422]: I have made this just a wrapper of a method in the sequence package now. To make things coherent I've done the same to `LocalRevision()`, as I think that we should reduce/remove usage of `SnapSequence.SideInfos()` as that method feels a bit like cheating, but let me know if this makes sense.

File: overlord/snapstate/snapstate_test.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate_test.go:5297]: Good point, added now

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the reviews, comments addressed now

**Review by pedronis — APPROVED** [review]
thanks, the new methods in sequence.go look good but probably now needs tests in the package?

**Comment by alfonsosanchezbeato** [comment]
I've added some more tests to `sequence` now. Rebased & commits squashed now.

**Pull request merged**